torch/_inductor/codegen/triton.py

```diff
@@ -1626,6 +1638,56 @@ def reduction_hint(node):
         else:
             return node.node.data.reduction_hint
 
+    @staticmethod
+    def can_use_32bit_indexing(buffers: Iterable[ir.Buffer]):
+        int_max = torch.iinfo(torch.int32).max
+        buf_numel = [sympy_product(buf.get_size()) for buf in buffers]
+        buf_storage_sizes = [buf.get_layout().storage_size() for buf in buffers]
+        for numel, storage_size in zip(buf_numel, buf_storage_sizes):
+            if V.graph.sizevars.size_hint(numel) > int_max:
+                return False
```

Comment:
What's an example of an operation where the indexing operations overflow, and `numel > int_max` but not ` storage_size > int_max`?

Reply:
The case I'm concerned about is broadcasting into a reduction, e.g.

```python
@torch.compile
def fn(a, b):
    return torch.sum(a + b)
```

where `a` has shape `(1, s0)` and `b` has shape `(s0, 1)`.

The generated reduction loop looks like:
```python
for roffset in range(0, rnumel, RBLOCK):
    rindex = roffset + rbase
    rmask = rindex < rnumel
    r1 = (rindex // ks0)
    r0 = rindex % ks0
    tmp0 = tl.load(in_ptr0 + (r1), rmask, eviction_policy='evict_last', other=0)
    tmp2 = tl.load(in_ptr1 + (r0), rmask, eviction_policy='evict_last', other=0)
    tmp2 = tmp0 + tmp1
    _tmp3 = tl.where(rmask, _tmp3 + tmp2, _tmp3)

```

You can see that `rindex` has a maximum value of `s0*s0` even though each input buffer only stores `s0` elements. Although I just tested it out and it looks like the input doesn't actually have the expanded shape... So it might be better to use the kernel's numel instead.

Reply:
Ah, right, we don't create independent indices for these. Note that that happens in your example without the `torch.sum`  as well. Just doing `a.unsqueeze(1) + b` for vectors `a`, `b` gets you that issue. 
Now, when it comes to where to look at for the size, I think you're right, and looking at `xnumel, rnumel` may be the safer thing to do. The intuition here being that we know that we start reading garbage from the pointers when the masks are false, that is, when the computations go over `{xyz}mask` or `rmask`. Actually, perhaps via this reasoning, we just need to test that these are no larger than `int_max` and we should be good? Also, since we split reductions, perhaps we just need to check `xnumel`?

Reply:
The reduction is necessary so the output buffer isn't larger than `int_max`, otherwise the `storage_size` check is enough. 

>  Actually, perhaps via this reasoning, we just need to test that these are no larger than `int_max` and we should be good?

If by "these" you mean `xnumel` and `rnumel` then that doesn't hold for strided indexing calculations.

>  Also, since we split reductions, perhaps we just need to check `xnumel`?

This only holds if the splitting installs guards, which I don't think is correct.

Reply:
After an offline discussion, this looks fine. Let's leave a comment with a few points that were discussed, in particular why checking `xnumel * rnumel` and `storage_size` should be good enough.